Revisions to land bcs generation for peatlands (L4SM Version 6), river routing, irrigation, and interannually varying LAI, VISDF, NIRDF

smahanam opened the pull request:
This branch  created NLv4p land BCs data (/discover/nobackup/smahanam/bcs/NLv4p/). 
This PR also includes numerous updates pertaining to science developments since the last BCs data set (NLv4, June 2020).
This PR is zero-diff : 1) the model does not enter this utility directory, 2) verified diff against NLv4 for parameters outside of the current updates.

**Deletions**
sibalb.f (replaced with module_sibalb.F90)
zenith.f (no use)
irrg_model.F90 (this was a temporary file that was sort of never used)

**Additions (New files)**
module_sibalb.F90
module_irrig_params.F90
mkMODISrst2tiles.F90
mkIrrigTiles.F90
mkIAV_LAIalb.F90 
modis_mcd15a2h_forward.py

**Source Code Modifications:**
1) For River Routing Model implementation
CombineRasters.F90
rasterize.F90
mkSMAPTilesPara_v2.F90
mkLandRaster.F90
make_bcs

2) For Irrigation model implementation
module_irrig_params.F90 **(NEW)**
mkIrrigTiles.F90 **(NEW)** 
mkCatchParam.F90

3) MODIS V6 MCD14A2H LAI and MCD43D VISDF/NIRDF preprocessors for MERRA3
Stage 1 preprocessor: modis_mcd15a2h_forward.py  **(NEW)**
Stage 2 preprocessor: mkMODISrst2tiles.F90 **(NEW)**

4) A program to create AGCM input files in MAPL_ReadForcing format for interannually varying LAI and VISDF/NIRDF scale parameters 
mkIAV_LAIalb.F90  **(NEW)**

5) PEATMAP
mod_process_hres_data.F90

6) EASE til file header changed from the LDASsa til file reader's format to GEOSldas (MAPL)
rmTinyCatchParaMod.F90
mkSMAPTilesPara_v2.F90
mkSMAPTilesPara.F90
mod_process_hres_data.F90

7) increased N_tiles_per_cell from 8 to 9 for CF5760
8) Some modifications to the subroutine that derives country/state boundaries
9) module_sibalb.F90 : old F77 subroutine was replaced by a F90 module and verified
10) to deal with SLES12 memory restrictions etc., make_bcs now submit jobs.
11) plottinng and README:  some changes related to new science developments (clsm_plots.pro, create_README.csh)

**Changes to BCs data files**
1) peat CLSM: Hybrid of HWSD peat and PEATMAP peat with peat CLSM parameters: soil_param.dat, ar.new, bf.dat, ts.dat, tau_param.dat and catch_params.nc4
2) River routing model
-Added 291,284 (number of Pfafstetter catchments globally) as the 2nd integer in the first line of master til file (and every til file)
-corrected to the 10th column in cubed-sphere master til files for proper representation of the pfafstetter grid
-introduced a supplementary til file for EASE grids by overlaying the EASE grid on pfafstetter catchments. This til file will be used  to transfer runoff from 1-tile-per-EASE-grid-cell space to catchment space
3) irrigation_IMxJM_DL.dat: new irrigation model parameters

gmao-rreichle commented:
Thanks, @smahanam , for creating the draft PR. 

This PR **only** changes the bcs generation package in ./Utils/Raster/

> 2) verified diff against NLv4 for parameters outside of the current updates.

@smahanam : Does this mean that the revised bcs generation package can still generate binary identical "NLv4" bcs (used in current SMAP L4SM Version 5 ops)?
And did you verify that the revised code can also still generate binary identical versions of the full suite of "NLv3" bcs that are currently used with the GCM (incl. FP)?   
"NLv3" must have been generated under SLES11.  I'm not sure if the bcs would be 0-diff under SLES12.  (I don't know whether "NLv4" might have been generated under SLES11 or SLES12.)

smahanam commented:
> Thanks, @smahanam , for creating the draft PR.
> 
> This PR **only** changes the bcs generation package in ./Utils/Raster/
> 
> > 1. verified diff against NLv4 for parameters outside of the current updates.
> 
> @smahanam : Does this mean that the revised bcs generation package can still generate binary identical "NLv4" bcs (used in current SMAP L4SM Version 5 ops)?
> And did you verify that the revised code can also still generate binary identical versions of the full suite of "NLv3" bcs that are currently used with the GCM (incl. FP)?
> "NLv3" must have been generated under SLES11. I'm not sure if the bcs would be 0-diff under SLES12. (I don't know whether "NLv4" might have been generated under SLES11 or SLES12.)

Inclusion of peat processing affects soil, and in return, ar*, bf* ts* and tau* parameters. So I disabled peat processing (have a .true./.false. switch somewhere in the code), and created a M36 BCs set - this is supposed to be identical to NLv4

/discover/nobackup/smahanam/bcs/NLv4/SMAP_EASEv2_M36/
 and compared against
 /discover/nobackup/projects/gmao/smap/bcs_temporary/NLv4/SMAP_EASEv2_M36/
and found ar*, bf* and ts* do differ while soil and topo files are identical. 
This is why we recommend using the same tag to reproduce params

visdf_IMxJM, and nirdf_IMxJM (used F90ized subalb) are zero-diff.

smahanam commented:
**An update !**
Added a runtime option to make_bcs to select a land BCs version from F25, GM4, ICA, NL3, NL4, or NL4p to derive a BCs set using the current version source codes. This option will not produce identical BCs to the archived BCs due to compiler differences, bug fixes and code improvements that have taken place since those BCs were created.  Science will match, however.  @yvikhlya , @sanAkel and @amolod requested this option. One last step needs to be added to fully automate tripolar grids.

All calculations in one subroutine were switched to double precision mode and replaced NRs F77 version of log gamma with NRs F90 as an attempt to address this non-zerodiff floating issue. Tests are still underway.

smahanam commented:
@gmao-rreichle Now that FP-IT also needs a long term BCs set scientifically equivalent to NL3set, and L4SM also awaits NLv4p, lets open this PR. 
I am sure @sdrabenh will also like to deal with one PR at a time.

gmao-rreichle commented:
@smahanam @sdrabenh 
I don't think we're ready for Scott's review of this PR.  It has yet to be reviewed by the land group. 
I do not know what the plans are for which tag will be used by GEOS-IT, but I suspect it will be either the CVS tag or the git tag for the GEOS 5.27.1 DAS, and not something that would include this PR.  
@rdkoster @wmputman @biljanaorescanin

smahanam commented:
OK I am closing this PR now. 
After creating NLv4p, my intention of this DRAFT for an internal review by the land group.
SLES12 related comment came, it was too important to brush aside, so I addressed it by changing a whole subroutine to double precision and replacing NRs 77 log gamma with NRs 90. And I deleted NL4p and recreated.

OK I am closing, and switching this to draft PR again.

smahanam commented:
It looks like I have to reopen as a PR, I can't switch to draft
@sdrabenh I am opening this but absolutely there is no rush. For now, just leave it, please.

mathomp4 commented:
@smahanam To convert to draft go up to the top of the PR under the Reviewers list. You'll see "Still in progress? Convert to draft". For some reason GitHub decides to make every link on this page blue...except for the one that "Convert to draft" is. I might ask about that...

smahanam commented:
> @smahanam To convert to draft go up to the top of the PR under the Reviewers list. You'll see "Still in progress? Convert to draft". For some reason GitHub decides to make every link on this page blue...except for the one that "Convert to draft" is. I might ask about that...

Thank you Matt !

smahanam commented:
Rolf, you may have noticed some activities in this branch last few days. Yury and Bin came together to add tripolar BCs to make_bcs. We finished those modifications. Also, they have nothing to with land modules so shouldn't have disturbed the internal review.

Added the below user guide to the GEOSgcm_GridComp wiki:
https://github.com/GEOS-ESM/GEOSgcm_GridComp/wiki/make_bcs-User-Guide

biljanaorescanin commented:
Is it possible for Bin or Yuri to push these updates needed for Ocean Tripolar on a separate PR from Sarith's?
My concern here is that we need more time for Sarith’s PR. If these changes are needed for BCS generation for Ocean as Santha said hotfix, I suppose you will want to have this as soon as possible. 
I also think Scott will appreciate smaller PR’s from all our sides. 
Is there any reasoning why this has to be on Sarith’s PR?

yvikhlya commented:
> Is it possible for Bin or Yuri to push these updates needed for Ocean Tripolar on a separate PR from Sarith's?
> My concern here is that we need more time for Sarith’s PR. If these changes are needed for BCS generation for Ocean as Santha said hotfix, I suppose you will want to have this as soon as possible.
> I also think Scott will appreciate smaller PR’s from all our sides.
> Is there any reasoning why this has to be on Sarith’s PR?

I am personally fine with using this branch for generating a coupled grids with new land BCs. No need to wait until it will be merged with develop.

gmao-rreichle commented:
@yvikhlya: I would advise against simply using this branch (or a commit from this branch).  The branch associated with this PR is likely to evolve considerably.  It may become obsolete because we need to break down the changes into smaller pieces that are contained in separate branches.  If you want to go this route, you need to make a copy of this branch for yourself and keep it safe.  
cc: @biljanaorescanin @rdkoster

yvikhlya commented:
> @yvikhlya: I would advise against simply using this branch (or a commit from this branch). The branch associated with this PR is likely to evolve considerably. It may become obsolete because we need to break down the changes into smaller pieces that are contained in separate branches. If you want to go this route, you need to make a copy of this branch for yourself and keep it safe.
> cc: @biljanaorescanin @rdkoster

Whatever branch has working make_bcs utility which is able to generate a working coupled set up with new land BCs is good for me. You guys decide where do you want to keep it and let us know (me, Bin and Santha).

smahanam commented:
Rolf, 

I am directly quoting you from SSAI's  mid-year performance review this morning, "The cited issues were: (1) bundled changes were too big to review all at once and (2) multiple components depend on the BC changes.”

OK I heard it loud and clear,  we split this. 
It will be me who is going to split this branch (Yury or Bin can't do it). 
Now Rolf, if we are to break this what to do you want to review first? I asked this very question in March. it has been 2+ months, nobody even asked for a walk through. With all due respect, I don't quite understand how you and Biljana review this by just diffing  when you know very little about the workflow in the first place. I guess this is your first time in this directory too.

Note this too - the programs in Raster serve multiple modules and GCs - not just land science. Other 2 draft PRs that I opened with this  draft PR are subject specific.  In fact, the most important changes in this branch are not even land science updates.
 
If you have figured out the way how this should be spitted, I would just follow it. Or leave it to me, I will break this branch to 3.
1) Urgent + miscellaneous
2) peatCLSM 
3) Irrig

in that order.

This needs a walk through before making any decision.
Alternatively, now that S2S tripolar issue has been resolved, we wait until I finish the river routing model. I can break this to 10 pieces and send a piece at a time.
